### PR TITLE
UI HexViewer

### PR DESCRIPTION
**Pull request by Kryptos-FR**
**Description**
We had an hex viewer implemented with `ImGui`. Replace with a similar one implemented with `egui`.

**Test plan**
- [x] should have the same features

**Usage**
*See sample*

### CODE REVIEW

File: rust/src/shards/gui/widgets/hex_viewer.rs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        std::slice::from_raw_parts_mut(data, len)
+      };
+
+      let mut mem_editor = MemoryEditor::new().with_address_range("All", 0..mem.len());
+
+      if let Some(options) = &self.editor_options {
+        mem_editor = mem_editor.with_options(options.clone());
+      }
+      mem_editor.draw_editor_contents_read_only(ui, mem, |mem, address| mem[address].into());
+      self.editor_options = Some(mem_editor.options.clone());
```
Comment (guusw) [inline rust/src/shards/gui/widgets/hex_viewer.rs:185]: I think you can keep the MemoryEditor instance on this shard, that way you also don't have to get/restore the options every frame.

Comment (guusw) [reply rust/src/shards/gui/widgets/hex_viewer.rs:185]: Just update the address range potentially

**Review by guusw — CHANGES_REQUESTED** [review]

File: rust/Cargo.toml [hunk 2]
```diff
@@ -88,9 +89,11 @@ shards = ["reqwest",
           "egui",
           "egui-gfx",
           "code_editor",
+          "hex_viewer",
           "tract-onnx",
           "enum-map"]
 code_editor = ["syntect"]
+hex_viewer = ["egui_memory_editor"]
```
Comment (sinkingsugar) [inline rust/Cargo.toml:96]: is it building by default?

File: rust/src/shards/gui/widgets/hex_viewer.rs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+          SHType_String => (
+            input.payload.__bindgen_anon_1.__bindgen_anon_2.stringValue as *mut u8,
+            input.payload.__bindgen_anon_1.__bindgen_anon_2.stringLen as usize,
+          ),
+          _ => unreachable!(),
+        };
+        std::slice::from_raw_parts_mut(data, len)
+      };
+
+      let mut mem_editor = MemoryEditor::new().with_address_range("All", 0..mem.len());
```
Comment (sinkingsugar) [inline rust/src/shards/gui/widgets/hex_viewer.rs:179]: anything new is hyper misleading, this is just the way UI api works (badly) right? doesn't mean heavy allocations...

File: rust/src/shards/gui/widgets/hex_viewer.rs (hunk 1 above)
Comment (sinkingsugar) [reply rust/src/shards/gui/widgets/hex_viewer.rs:185]: yeah.... I think that needs to be done

**Review by sinkingsugar — CHANGES_REQUESTED** [review]

File: rust/src/shards/gui/widgets/hex_viewer.rs (hunk 1 above)
Comment (Kryptos-FR) [reply rust/src/shards/gui/widgets/hex_viewer.rs:185]: Unfortunately it can't be done. The address range is not writeable, except during construction.

Comment (sinkingsugar) [reply rust/src/shards/gui/widgets/hex_viewer.rs:185]: I mean all the `new` around egui

File: rust/src/shards/gui/widgets/hex_viewer.rs (hunk 3 above)
Comment (Kryptos-FR) [reply rust/src/shards/gui/widgets/hex_viewer.rs:179]: `with_address_range` only available during construction: it moves `Self` (doesn't return a reference), so if we keep a reference to the `MemoryEditor`, we can't modify the range afterwards. It's a limitation of this library currently.

File: rust/Cargo.toml (hunk 2 above)
Comment (Kryptos-FR) [reply rust/Cargo.toml:96]: Yes.

File: rust/src/shards/gui/widgets/hex_viewer.rs (hunk 3 above)
Comment (Kryptos-FR) [reply rust/src/shards/gui/widgets/hex_viewer.rs:179]: Issue opened: https://github.com/Hirtol/egui_memory_editor/issues/5

Comment (Kryptos-FR) [reply rust/src/shards/gui/widgets/hex_viewer.rs]: PR created: Hirtol/egui_memory_editor#6

File: rust/src/shards/gui/widgets/hex_viewer.rs (hunk 1 above)
Comment (sinkingsugar) [reply rust/src/shards/gui/widgets/hex_viewer.rs]: I think we discussed this on discord

**Review by sinkingsugar — APPROVED** [review]

**Pull request merged**